### PR TITLE
podman pod top

### PR DESCRIPTION
**Pull request by haircommander**
Signed-off-by: haircommander <pehunt@redhat.com>

### CODE REVIEW

**Comment by haircommander** [comment]
Added WIP because there are some inefficiencies that could be fixed.
1: I had to copy functions from cmd/podman/shared to find the pid namespace information for the containers. This was the best way to filter duplicates (containers in the same namespace would return the same output twice). I don't know of a solution to this problem.
2: The psgo library doesn't have an option to run without a header, so I had to filter it out manually. @vrothberg is this something you'd be willing to add? If so, I can take a stab at it, but i'm sure you'd handle it much faster if you have time.
Everything else I feel pretty good about.

**Comment by haircommander** [comment]
@mheon @baude @rhatdan PTAL

**Comment by rhatdan** [comment]
psgo should definitely take a flag to show no header.

**Comment by vrothberg** [comment]
I am against adding a no-header parameter. Oftentimes, adding more parameters is adding complexity at the users' side that should be handled by the library. The above issue is an example for such a case.

My suggestion is to instead add another API to extract the data for a set of processes as proposed by @marcov here https://github.com/containers/psgo/issues/32. All the logic can be implemented in `psgo` and libpod only needs to pass a slice of pids.

@haircommander, if you like this proposal, we can extend psgo early next week and get everything into the next Podman release on time.

**Comment by haircommander** [comment]
@vrothberg Yup SGTM

**Comment by rhatdan** [comment]
@vrothberg SGTM

**Comment by vrothberg** [comment]
Cool, we'll ping you in the PR for containers/psgo to have a look and provide feedback.

**Comment by haircommander** [comment]
@vrothberg's update fixed both of my issues. Should be good to go. PTAL

**Comment by vrothberg** [comment]
@haircommander awesome! There is just one issue with cross compilation:
```
TARGET="darwin.amd64"; \
	GOOS="${TARGET%%.*}" \
	GOARCH="${TARGET##*.}" \
	go build -i -ldflags ' -X main.gitCommit="9db0435d7da50e35f925e77055611712983bc4f8" -X main.buildInfo=1534973108' -tags 'containers_image_openpgp containers_image_ostree_stub exclude_graphdriver_btrfs exclude_graphdriver_devicemapper exclude_graphdriver_overlay' -o "bin/podman.cross.darwin.amd64" github.com/containers/libpod/cmd/podman
# github.com/containers/libpod/cmd/podman
cmd/podman/pod_top.go:90:22: pod.GetPodPidInformation undefined (type *libpod.Pod has no field or method GetPodPidInformation)
```

**Comment by vrothberg** [comment]
We need something like `libpod/pod_top_unsupported.go` to export the stubs to make the builds happy.

File: cmd/podman/pod_top.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "unable to lookup requested container")
+	}
+
+	podStatus, err := shared.GetPodStatus(pod)
+	if err != nil {
+		return err
+	}
+	if podStatus != "Running" {
+		return errors.Errorf("pod top can only be used on pods with running containers")
```
Comment (vrothberg) [inline cmd/podman/pod_top.go]: I propose to use "with at least one running container"

File: test/e2e/pod_top_test.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		session.WaitWithDefaultTimeout()
+		Expect(session.ExitCode()).To(Equal(0))
+
+		result := podmanTest.Podman([]string{"pod", "top", podid})
+		result.WaitWithDefaultTimeout()
+		Expect(result.ExitCode()).To(Equal(0))
+		Expect(len(result.OutputToStringArray())).To(Equal(4))
+	})
+
+	It("podman pod top on pod with containers in same different namespace", func() {
```
Comment (vrothberg) [inline test/e2e/pod_top_test.go]: remove "same"

**Comment by vrothberg** [comment]
It might seem like bikeshedding, but I really value expressive commit messages. They're invaluable when doing research and help a lot when debugging etc. Can you add some descriptions to the commits?

Despite these little nitpicks, code LGTM.

File: libpod/pod_top_linux.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+// GetPodPidInformation returns process-related data of all processes in
+// the pod.  The output data can be controlled via the `descriptors`
+// argument which expects format descriptors and supports all AIXformat
+// descriptors of ps (1) plus some additional ones to for instance inspect the
+// set of effective capabilities.  Eeach element in the returned string slice
+// is a tab-separated string.
+//
+// For more details, please refer to github.com/containers/psgo.
+func (p *Pod) GetPodPidInformation(descriptors []string) ([]string, error) {
```
Comment (mheon) [inline libpod/pod_top_linux.go]: Should probably lock the pod during this function, to ensure that it isn't removed out from under you

File: libpod/pod_top_unsupported.go [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// +build !linux
+
+package libpod
+
+// GetPodPidInformation is exclusive to linux
+func (p *Pod) GetPodPidInformation(descriptors []string) ([]string, error) {
+    return nil, nil
```
Comment (mheon) [inline libpod/pod_top_unsupported.go]: Probably return ErrNotImplemented here

File: libpod/pod_top_linux.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	pids := make([]string, 0)
+	ctrsInPod, err := p.AllContainers()
+	if err != nil {
+		return nil, err
+	}
+	for _, c := range ctrsInPod {
+		c.lock.Lock()
```
[the remark this answers is not in the document]
Comment (haircommander) [reply libpod/pod_top_linux.go]: I've grown lazy with the option of defers... good catch

File: libpod/pod_top_linux.go (hunk 3 above)
Comment (haircommander) [reply libpod/pod_top_linux.go]: fixed!

**Comment by mheon** [comment]
LGTM. @rh-atomic-bot r+

**Pull request closed without merging**